### ISSUE TITLE
V6: Use better-sqlite3

### ISSUE THREAD

**Issue by sushantdhiman** [sequelize/sequelize#11400]
Thinking about using `better-sqlite3` for SQLite database, mainly for supporting 64bit Integers. Their benchmark looks great as well

https://github.com/JoshuaWise/better-sqlite3/issues/262

**Comment by papb** [comment]
Completely agreed!!

**Comment by usebaz** [comment]
@sushantdhiman Hi, if need help I can make PR with migrate to `better-sqlite3`

**Comment by sushantdhiman** [comment]
@usebaz Thanks. Soon `master` branch will point to the next beta. After that you can submit your PR

**Comment by keshav-bohr** [comment]
@sushantdhiman 
Just a thought. We are adding too many database support in `sequelize` and that is great.
But can we have something like separate module for every db, that could reduce the load.

There would be no search for dialect every time. It might optimise the performance. Not drastically, but to some extend.
What are your thoughts about this?

**Comment by etylermoss** [comment]
Is there any update on this?

**Comment by sushantdhiman** [comment]
@etylermoss Currently no one is working on this. I tried to convert some time ago but found that API is really annoying :)

For example,

1) [`run`](https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md#runbindparameters---object), `(only on statements that do not return data)`
2) [`get`](https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md#getbindparameters---row) `(only on statements that return data)`

This mean as an ORM we need to check what kind of query is going to execute (will it return data?) and switch methods based on that. I think this is unnecessary complication.

I wish they could just provide simple method to execute any arbitrary `query` (with parameters) like every other dialect library in existence.

So officially we want to switch but we are not working on it. If someone want to do a PR for this, we will welcome that change.

**Comment by louy2** [comment]
How is it handled with `node-sqlite3` currently? It has the same run/get split.

**Comment by meshakeeb** [comment]
Are you going to add this anytime soon?

**Comment by lesion** [comment]
> @etylermoss Currently no one is working on this. I tried to convert some time ago but found that API is really annoying :)
> For example,
>     1. [`run`](https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md#runbindparameters---object), `(only on statements that do not return data)`
> 
>     2. [`get`](https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/api.md#getbindparameters---row) `(only on statements that return data)`

> This mean as an ORM we need to check what kind of query is going to execute (will it return data?) and switch methods based on that. I think this is unnecessary complication.
> 
> I wish they could just provide simple method to execute any arbitrary `query` (with parameters) like every other dialect library in existence.
> 
> So officially we want to switch but we are not working on it. If someone want to do a PR for this, we will welcome that change.

Hi there!
Is your attempt available somewhere? I'm thinking of trying to implement this

**Comment by bradisbell** [comment]
Not stale.

**Comment by John-H-Smith** [comment]
Any Updates?

**Comment by ephys** [comment]
@John-H-Smith we're thinking about doing it for v7 https://github.com/sequelize/meetings/issues/7

**Comment by cirosantilli** [comment]
I reproduce their 10x faster benchmarks in my project when migrating from better-sqlite3 to sequelize 5, It is amazing!!

Make sure to compile with that log limiting options maxed out if possible to prevent log truncation: https://github.com/JoshuaWise/better-sqlite3/issues/306

**Comment by tankakatan** [comment]
Hi everyone,

Please consider adding support for `better-sqlite3` package ASAP. We are experiencing lots of issues with `node-sqlite3` when shipping our Node.js app as a packaged executable for different platforms. Even though it is possible to include it in the bundle, our users keep reporting very weird and hard to debug issues with the old lib.

**Comment by ephys** [comment]
At this moment none of the maintainers are working on replacing sqlite3 with better-sqlite3 and won't be in the foreseeable future since there are many other features that we want to add, and bugs we need to fix. If somebody from the community makes a Pull Request we are of course happy to review that.

**Comment by tankakatan** [comment]
Thank you very much for the quick response @ephys. In our team we'll estimate the effort needed for the integration, and find out whether we are able to allocate resources for that. Also pinging @usebaz who [offered](https://github.com/sequelize/sequelize/issues/11400#issuecomment-542109003) his help with the integration a few years ago. Just in hope he is still available.

**Comment by WikiRik** [comment]
No updates. We've moved this internally to v8 when we split the various dialects off to separate packages. That way we can even look into supporting multiple connectors for sqlite

**Comment by dixiekong** [comment]
any updates?

**Comment by j-refs** [comment]
Is this still not being worked on? If the maintainers approve, I'll gladly open a fork open to the community to get this feature up and running and PR it.

After a quick scan of the source, it doesn't really seem as hard as every is making it, seems like the regular `sqlite` dialect can be easily modified for `better-sqlite3`, but I just don't have the time to scan and port everything, maybe if we make it a community effort, we can get this done

**Comment by WikiRik** [comment]
@j-refs internally this has been postponed to v8 when we split dialects. If you can make it so that users can select which of the two adapters they want to use, we can reconsider adding this to v7.

If you create a fork, please mention it here in this issue as well so we can refer people to it in the meantime

**Comment by j-refs** [comment]
Will look into forsure. Probably later today, I work nights and I'm ready to KO. Any info the maintainers wanna add in before I form?

**Comment by yisar** [comment]
any updates?

**Comment by j-refs** [comment]
Yes sorry, struggling to find time to get started on this, but is definitely on my queue

**Comment by j-refs** [comment]
I'm ready to start working on this. Would y'all prefer a branched PR or it's own repo?

**Comment by ephys** [comment]
@j-refs So sorry I did not see your latest message

The SQLite package has been split to its own package (https://github.com/sequelize/sequelize/tree/main/packages). You can fork the repository, create a new package called `better-sqlite3`, and open a PR to this repository :)

**Comment by spuxx1701** [comment]
I'm in dire need of better-sqlite3 support. Is this being worked on or nah?

**Comment by 0xslipk** [comment]
@ephys I will start working on this. Do you guys have something already in progress that can work or help with?

**Comment by j-refs** [comment]
> [@j-refs](https://github.com/j-refs) So sorry, I did not see your latest message
> 
> The SQLite package has been split to its own package (https://github.com/sequelize/sequelize/tree/main/packages). You can fork the repository, create a new package called `better-sqlite3`, and open a PR to this repository :)
> 
> As we're opting for providing both `@sequelize/sqlite3` and `@sequelize/better-sqlite3`, this won't be a breaking change after all
> 
> Note that `@sequelize/better-sqlite3` cannot extend `@sequelize/sqlite3` directly to avoid installing both sqlite3 and better-sqlite3. Instead, anything that you wish to re-use from `@sequelize/sqlite3` should be moved to an internal package, `@sequelize/abstract-sqlite`

I was not able to put any effort into this sorry. I ran into some major medical issues and am kind of barely getting backaround

**Comment by Steve-Mcl** [comment]
FYI: sqlite3 now officially deprecated.

<img width="922" height="384" alt="Image" src="https://github.com/user-attachments/assets/78659738-e071-4bd1-b6a8-d9f773610156" />

https://github.com/TryGhost/node-sqlite3/commit/a85f9e880aa065ef7a6ff3a8a555b0ed2c5015a1

Wondering if anyone is using a fork which doesnt have the TAR CVE
```
tar  <=7.5.3
Severity: high
node-tar is Vulnerable to Arbitrary File Overwrite and Symlink Poisoning via Insufficient Path Sanitization - https://github.com/advisories/GHSA-8qq5-rm4j-mr97
Race Condition in node-tar Path Reservations via Unicode Ligature Collisions on macOS APFS - https://github.com/advisories/GHSA-r6q2-hw4h-h46w
```

```
$ npm ls tar
blahblah
└─┬ sqlite3@5.1.7
  ├─┬ node-gyp@8.4.1
  │ ├─┬ make-fetch-happen@9.1.0
  │ │ └─┬ cacache@15.3.0
  │ │   └── tar@6.2.1 deduped
  │ └── tar@6.2.1 deduped
  └── tar@6.2.1
```

**Comment by HoldYourWaffle** [comment]
Worth noting that the first-party `node:sqlite` module [moved to release candidate](https://github.com/nodejs/node/pull/61262) in Node v25.7.
